flow/aggregates.py

```diff
@@ -0,0 +1,463 @@
+        """
+        aggregated_jobs = self._generate_aggregates(project)
+        self._create_nested_aggregate_list(aggregated_jobs, project)
+
+    def _generate_aggregates(self, project):
+        jobs = project
+        if self._aggregator._select is not None:
+            jobs = filter(self._aggregator._select, jobs)
+        if self._aggregator._sort_by is None:
+            jobs = list(jobs)
+        else:
+            jobs = sorted(jobs,
+                          key=lambda job: job.sp[self._aggregator._sort_by],
+                          reverse=self._aggregator._reverse_order)
+        yield from self._aggregator._aggregator_function(jobs)
+
+    def _create_nested_aggregate_list(self, aggregated_jobs, project):
+        """signac-flow internally assumes every aggregate to be a tuple of jobs.
+
+        This method converts every aggregate in ``aggregated_jobs``, which may be of
+        any type, returned from an aggregator_function using the instance of
+        ``_MakeAggregates`` to an aggregate of jobs as tuple.
+        """
+        def _validate_and_filter_job(job):
+            "Validate whether a job is eligible to be a part of an aggregate or not."
+            if job is None:
+                return False
+            elif job in project:
+                return True
+            else:
+                raise LookupError(f'The signac job {job.get_id()} not found'
+                                  f'in {project}')
+
+        for aggregate in aggregated_jobs:
+            try:
+                filter_aggregate = tuple(filter(_validate_and_filter_job, aggregate))
+            except TypeError:  # aggregate is not iterable
+                ValueError("Invalid aggregator_function provided by the user.")
+            # Store aggregate in this instance
+            self._aggregates.append(filter_aggregate)
+            # Store aggregate by their ids in order to search through id
+            self._aggregate_ids[get_aggregate_id(filter_aggregate)] = filter_aggregate
+
+
+class _DefaultAggregateStore:
+    """This class holds the information of the project associated with
+    an operation function aggregated by the default aggregates i.e.
+    :py:class:`aggregator.groupsof(1)`.
+
+    When iterated through it's instance, it yields a tuple of a single job from
+    the Project.
+
+    :param project:
+        A signac project used to fetch jobs for creating aggregates.
+    :type project:
+        :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+    """
+    def __init__(self, project):
+        self._project = project
+
+    def __iter__(self):
+        for job in self._project:
+            yield (job,)
+
+    def __getitem__(self, id):
+        "Return a tuple of a single job via job id."
+        try:
+            return (self._project.open_job(id=id),)
+        except KeyError:
+            raise LookupError(f"Did not find job with id {id}.")
```

Comment:
Should this error be about a job or an aggregate? This is a forward-looking question about how the aggregate's contains method will actually be invoked, @b-butler in future work making use of this class which error would be more appropriate? It seems like we'd want the error to state something about aggregates, but since this class will become the default mode for all aggregation maybe that's not the case since this error could get thrown for operations that act on individual jobs.

Reply:
That is true.
In #335 I have added support for these classes and I believe that it should be about an aggregate because everything will become an aggregate internally.
I'm curious to know what @b-butler thinks about this